### PR TITLE
curvefs/tool: support query inode and list partition

### LINKED ISSUE #1218: curvefs/tool: query inode
**Is your feature request related to a problem? (你需要的功能是否与某个问题有关?)**

query inode by curvefs::metaserver::MetaServerService_Stub::GetOrModifyS3ChunkInfo

**Describe the solution you'd like (描述你期望的解决方法)**

**Describe alternatives you've considered (描述你想到的折衷方案)**

**Additional context/screenshots (更多上下文/截图)**

### PR DESCRIPTION
**Pull request by Cyber-SiKu**
### What problem does this PR solve?

Issue Number: close #1218  

Problem Summary:

### What is changed and how it works?

What's Changed:

1. use GetOrModifyS3ChunkInfoResponse to query inode
2. add list-partition by fsId for locating inodes
3. use GeIndode to get inode baseInfo

How it Works:
```
curvefs_tool   query-inode -fsId=1 -inodeId=167772167
fsId: 1 inodeId: 16777217
-baseInfo:
inodeId: 16777217
fsId: 1
length: 3073219
ctime: 1651219921
ctime_ns: 647724164
mtime: 1651219921
mtime_ns: 647724164
atime: 1651219921
atime_ns: 647724164
uid: 13011
gid: 3000
mode: 33188
nlink: 1
type: TYPE_S3
rdev: 0
s3ChunkInfoMap {
  key: 0
  value {
    s3Chunks {
      chunkId: 2
      compaction: 0
      offset: 0
      len: 3073219
      size: 3073219
      zero: false
    }
  }
}
openmpcount: 0
parent: 1
-----------------
-s3ChunkInfo:
chunkId: 2
compaction: 0
offset: 0
len: 3073219
size: 3073219
zero: false

Total s3chunkinfo: 1
```

Side effects(Breaking backward compatibility? Performance regression?): 

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: curvefs/src/tools/query/curvefs_inode_query.cpp [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+namespace query {
+
+void InodeQueryTool::PrintHelp() {
+    CurvefsToolRpc::PrintHelp();
+    std::cout << " -poolId=" << FLAGS_poolId
+              << " -copysetId=" << FLAGS_copysetId
+              << " -partitionId=" << FLAGS_partitionId
+              << " -fsId=" << FLAGS_fsId << " -inodeId=" << FLAGS_inodeId
+              << " [-metaserverAddr=" << FLAGS_metaserverAddr << "]";
+    std::cout << std::endl;
```
Comment (wu-hanqing) [inline curvefs/src/tools/query/curvefs_inode_query.cpp:45]: I think it's your responsibility to find inode's pool/copyset/partition.

File: curvefs/src/tools/query/curvefs_inode_query.cpp [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    if (poolsId.size() != copysetsId.size() ||
+        poolsId.size() != partitionId.size() || poolsId.size() != fsId.size() ||
+        poolsId.size() != inodeId.size()) {
+        std::cout << "poolId, copysetId, partitionId, fsId, inodeId must be "
+                     "the same size"
+                  << std::endl;
+        return -1;
+    }
+
+    for (unsigned i = 0; i < poolsId.size(); ++i) {
```
Comment (wu-hanqing) [inline curvefs/src/tools/query/curvefs_inode_query.cpp:83]: `size_t i = 0`

File: curvefs/src/tools/query/curvefs_inode_query.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+namespace curvefs {
+namespace tools {
+namespace query {
+
+using HostAndResponseType =
+    std::vector<std::pair<std::string,
+                          curvefs::metaserver::GetOrModifyS3ChunkInfoResponse>>;
+
+struct CompareRequest {
+    bool operator()(
```
Comment (wu-hanqing) [inline curvefs/src/tools/query/curvefs_inode_query.h:47]: `bool operator() const`

File: curvefs/src/tools/query/curvefs_inode_query.h [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        return a.poolid() == b.poolid() && a.copysetid() == b.copysetid() &&
+               a.partitionid() == b.partitionid() && a.fsid() == b.fsid() &&
+               a.inodeid() == b.inodeid();
+    }
+};
+
+class InodeQueryTool
+    : public CurvefsToolRpc<curvefs::metaserver::GetOrModifyS3ChunkInfoRequest,
+                            curvefs::metaserver::GetOrModifyS3ChunkInfoResponse,
+                            curvefs::metaserver::MetaServerService_Stub> {
```
Comment (wu-hanqing) [inline curvefs/src/tools/query/curvefs_inode_query.h:59]: should we get inode's attributes ?

File: curvefs/src/tools/query/curvefs_inode_query.cpp [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        request.set_copysetid(std::stoi(copysetsId[i]));
+        request.set_partitionid(std::stoi(partitionId[i]));
+        request.set_fsid(std::stoi(fsId[i]));
+        request.set_inodeid(std::stoi(inodeId[i]));
+        request.set_returns3chunkinfomap(true);
+        AddRequest(request);
+    }
+
+    service_stub_func_ = std::bind(
+        &curvefs::metaserver::MetaServerService_Stub::GetOrModifyS3ChunkInfo,
```
Comment (SeanHai) [inline curvefs/src/tools/query/curvefs_inode_query.cpp:95]: If you want to get the whole s3infomap may be need using stream rpc api when it is too larger. And you can refer to the implementation in metaserver_client.cpp.

File: curvefs/src/tools/query/curvefs_inode_query.cpp (hunk 2 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/query/curvefs_inode_query.cpp:83]: > `size_t i = 0`

fix

File: curvefs/src/tools/query/curvefs_inode_query.cpp [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    std::vector<std::string> partitionId;
+    curve::common::SplitString(FLAGS_partitionId, ",", &partitionId);
+
+    std::vector<std::string> fsId;
+    curve::common::SplitString(FLAGS_fsId, ",", &fsId);
+
+    std::vector<std::string> inodeId;
+    curve::common::SplitString(FLAGS_inodeId, ",", &inodeId);
+
+    if (poolsId.size() != copysetsId.size() ||
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/tools/query/curvefs_inode_query.cpp:74]: > Does this means support query multiple inodes once, but the related info need one-to-one correspondence?

yep

**Comment by aspirer** [comment]
pls change your pr title.....

File: curvefs/src/tools/query/curvefs_inode_query.h (hunk 3 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/query/curvefs_inode_query.h:47]: > `bool operator() const`

fix

File: curvefs/src/tools/query/curvefs_inode_query.cpp (hunk 5 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/query/curvefs_inode_query.cpp:95]: > If you want to get the whole s3infomap may be need using stream rpc api when it is too larger. And you can refer to the implementation in metaserver_client.cpp.

fix

File: curvefs/src/tools/query/curvefs_inode_query.cpp (hunk 1 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/query/curvefs_inode_query.cpp]: > I think it's your responsibility to find inode's pool/copyset/partition.

fix

Comment (Cyber-SiKu) [reply curvefs/src/tools/query/curvefs_inode_query.cpp]: > agree, only fsId and inodeId are needed, and you can get other infos you need inside.

fix

File: curvefs/src/tools/query/curvefs_inode_query.h (hunk 4 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/query/curvefs_inode_query.h]: > should we get inode's attributes ?

fix

File: curvefs/src/tools/curvefs_tool.h [hunk 7]
```diff
@@ -120,13 +128,15 @@ class CurvefsToolRpc : public CurvefsTool {
              const std::shared_ptr<ResponseT>& response,
              const std::shared_ptr<ServiceT>& service_stub,
              const std::function<void(ControllerT*, RequestT*, ResponseT*)>&
-                 service_stub_func) {
+                 service_stub_func,
+            const std::shared_ptr<StreamClient> &streamClient) {
```
Comment (SeanHai) [inline curvefs/src/tools/curvefs_tool.h:132]: align

File: curvefs/src/tools/curvefs_tool.h [hunk 8]
```diff
@@ -196,32 +207,61 @@ class CurvefsToolRpc : public CurvefsTool {
...
      */
     virtual bool SendRequestToServices() {
-        uint32_t failHostNumner = 0;
+        uint32_t failHostNumber = 0;
+        bool ret = false;
         for (const std::string& host : hostsAddr_) {
             SetController();
-            if (channel_->Init(host.c_str(), nullptr) != 0) {
-                std::cerr << "fail init channel to host: " << host << std::endl;
+            std::shared_ptr<brpc::ChannelOptions> channelOpt = nullptr;
```
Comment (wu-hanqing) [inline curvefs/src/tools/curvefs_tool.h:214]: No need to allocate channel options on heap, a stack variable is enough.

File: curvefs/src/tools/curvefs_tool_define.h [hunk 9]
```diff
@@ -74,12 +74,16 @@ const char kPartitionQueryCmd[] = "query-partition";
...
 const char kFsQueryCmd[] = "query-fs";
+// Inode-query
+const char kInodeQueryCmd[] = "query-inode";
 // fsinfo-list
 const char kFsInfoListCmd[] = "list-fs";
 // list-fs-copysetid
 const char kCopysetInfoListCmd[] = "list-copysetInfo";
 // list-topology
 const char kTopologyListCmd[] = "list-topology";
+// list-topology
```
Comment (SeanHai) [inline curvefs/src/tools/curvefs_tool_define.h:85]: list-partition?

File: curvefs/src/tools/list/curvefs_partition_list.cpp [hunk 10]
```diff
@@ -0,0 +1,109 @@
+/*
+ *  Copyright (c) 2021 NetEase Inc.
```
Comment (SeanHai) [inline curvefs/src/tools/list/curvefs_partition_list.cpp:2]: 2022, check all files created this time.

File: curvefs/src/tools/query/curvefs_inode_s3infomap.cpp [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+
+void InodeS3InfoMapTool::PrintHelp() {
+    CurvefsToolRpc::PrintHelp();
+    std::cout << " -poolId=" << FLAGS_poolId
+              << " -copysetId=" << FLAGS_copysetId
+              << " -partitionId=" << FLAGS_partitionId
+              << " -fsId=" << FLAGS_fsId << " -inodeId=" << FLAGS_inodeId
+              << " [-metaserverAddr=" << FLAGS_metaserverAddr
+              << " -rpcStreamIdleTimeoutMs=" << FLAGS_rpcStreamIdleTimeoutMs
+              << "]";
```
Comment (SeanHai) [inline curvefs/src/tools/query/curvefs_inode_s3infomap.cpp:62]: Still need poolId copysetId and partitionId when query inode?

File: curvefs/src/tools/curvefs_tool.h [hunk 12]
```diff
@@ -196,32 +207,61 @@ class CurvefsToolRpc : public CurvefsTool {
...
+            if (isStreaming_) {
+                // set stream rpc client
+                channelOpt = std::make_shared<brpc::ChannelOptions>();
+                channelOpt->connection_group = "streaming";
+                StreamOptions streamOpt(FLAGS_rpcStreamIdleTimeoutMs);
+                connection_ = streamClient_->Connect(
+                    controller_.get(), receiveCallback_, streamOpt);
+                if (nullptr == connection_ || nullptr == receiveCallback_) {
```
Comment (wu-hanqing) [inline curvefs/src/tools/curvefs_tool.h:222]: should you check `receiveCallback_ == nullptr` before Connect or just after `if (isStreaming_)`?

File: curvefs/src/tools/query/curvefs_inode_s3infomap.h [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+struct HashInodeBase {
+    size_t operator()(const InodeBase& inode) const {
+        auto inodeIdHash = std::hash<uint64_t>()(inode.inodeid());
+        auto fsIdHash = std::hash<uint64_t>()(inode.fsid());
+        return inodeIdHash ^ fsIdHash;
+  }
+};
+
+struct KeyEuqalInodeBase {
+    size_t operator()(const InodeBase& a, const InodeBase& b) const {
```
Comment (wu-hanqing) [inline curvefs/src/tools/query/curvefs_inode_s3infomap.h:61]: return value is `size_t` or `bool`?

File: curvefs/src/tools/query/curvefs_inode_s3infomap.h [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+                          curvefs::metaserver::GetOrModifyS3ChunkInfoResponse>>;
+
+using InodeBase = curvefs::metaserver::GetInodeRequest;
+
+struct HashInodeBase {
+    size_t operator()(const InodeBase& inode) const {
+        auto inodeIdHash = std::hash<uint64_t>()(inode.inodeid());
+        auto fsIdHash = std::hash<uint64_t>()(inode.fsid());
+        return inodeIdHash ^ fsIdHash;
+  }
```
Comment (wu-hanqing) [inline curvefs/src/tools/query/curvefs_inode_s3infomap.h:57]: indentent

File: curvefs/src/tools/query/curvefs_inode_s3infomap.h [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+};
+
+struct KeyEuqalInodeBase {
+    size_t operator()(const InodeBase& a, const InodeBase& b) const {
+        return a.fsid() == b.fsid() && a.inodeid() == b.inodeid();
+    }
+};
+
+struct CompareRequest {
```
Comment (wu-hanqing) [inline curvefs/src/tools/query/curvefs_inode_s3infomap.h:66]: it seems `CompareRequest` is not used?

File: curvefs/src/tools/query/curvefs_inode_s3infomap.h [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+ protected:
+    void AddUpdateFlags() override;
+    bool AfterSendRequestToHost(const std::string& host) override;
+    bool CheckRequiredFlagDefault() override;
+    void SetReceiveCallback();
+    void UpdateInode2S3ChunkInfoList_(const InodeBase& inode,
+                                      const S3ChunkInfoList&list);
+
+ protected:
+    std::unordered_map<InodeBase, S3ChunkInfoList, HashInodeBase,
```
Comment (wu-hanqing) [inline curvefs/src/tools/query/curvefs_inode_s3infomap.h:104]: why the key is a rpc request ?

File: curvefs/src/tools/query/curvefs_inode_s3infomap.cpp [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+                     << ", error text " << controller_->ErrorText() << "\n";
+    } else {
+        if (response_->statuscode() == metaserver::MetaStatusCode::OK) {
+            if (!isStreaming_) {
+                if (response_->s3chunkinfomap_size() == 0) {
+                    UpdateInode2S3ChunkInfoList_(
+                        TraslateInodeBase(requestQueue_.front()),
+                        S3ChunkInfoList());
+                }
+                for (auto list : response_->s3chunkinfomap()) {
```
Comment (wu-hanqing) [inline curvefs/src/tools/query/curvefs_inode_s3infomap.cpp:140]: `for (const auto&...)`

File: curvefs/src/tools/query/curvefs_inode_s3infomap.cpp (hunk 11 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/query/curvefs_inode_s3infomap.cpp:62]: > Still need poolId copysetId and partitionId when query inode?

no just fsid and copysetId

File: curvefs/src/tools/curvefs_tool.h (hunk 7 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/curvefs_tool.h:132]: fix

File: curvefs/src/tools/list/curvefs_partition_list.cpp (hunk 10 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/list/curvefs_partition_list.cpp:2]: > 2022, check all files created this time.

fix

File: curvefs/src/tools/curvefs_tool.h (hunk 8 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/curvefs_tool.h:214]: > No need to allocate channel options on heap, a stack variable is enough.

fix

File: curvefs/src/tools/query/curvefs_inode_s3infomap.h (hunk 13 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/query/curvefs_inode_s3infomap.h:61]: > return value is `size_t` or `bool`?

fix

File: curvefs/src/tools/query/curvefs_inode_s3infomap.h (hunk 14 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/query/curvefs_inode_s3infomap.h:57]: > indentent

fix

File: curvefs/src/tools/query/curvefs_inode_s3infomap.h (hunk 15 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/query/curvefs_inode_s3infomap.h:66]: remove

File: curvefs/src/tools/query/curvefs_inode_s3infomap.cpp (hunk 17 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/query/curvefs_inode_s3infomap.cpp:140]: > `for (const auto&...)`
fix

File: curvefs/src/tools/query/curvefs_inode_s3infomap.h (hunk 16 above)
Comment (Cyber-SiKu) [reply curvefs/src/tools/query/curvefs_inode_s3infomap.h:93]: > why the key is a rpc request ?

The necessary information is the inode's fsId, poolid, copysetid, partitionid, and inodeid, and the above information can be provided by GetInodeRequest. If the request is used directly, the conversion can be reduced.

**Review by SeanHai — APPROVED** [review]

**Review by wu-hanqing — APPROVED** [review]

**Pull request merged**